### PR TITLE
Turn on web socket compression

### PR DESCRIPTION
**Pull request by swankjesse**
Theres a few TODOs outstanding around error cases, and I need to
run integration tests against real-world things. But the core
functionality works.

I used coverage to confirm the context takeover cases are exercised
with and without.

### CODE REVIEW

**Comment by swankjesse** [comment]
I wanna do an OkHttp 4.5 'Release Candidate' with this feature so early adopters can check whether this works before everyone else does.

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt [hunk 1]
```diff
@@ -607,5 +637,15 @@ class RealWebSocket(
...
+
+    /**
+     * The smallest message that will be compressed. We use 1024 because smaller messages already
+     * fit comfortably within a single ethernet packet (1500 bytes) even with framing overhead.
+     *
+     * For tests this must be big enough to realize real compression on test messages like
+     * 'aaaaaaaaaa...'. Our tests check if compression was applied just by looking at the size if
+     * the inbound buffer.
+     */
+    const val DEFAULT_MINIMUM_DEFLATE_SIZE = 1024L
```
Comment (AntonRutkevich) [inline okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: @swankjesse , consider the following case: 

An app is sending a stream of coordinates, JSON encoded. It has dozen of fields (lat, lon, accuracy, heading, timestamp, ...), so the size of the uncompressed JSON is ~110..130B.
With context takeover, a stream of such messages gets compressed quite well (~10-20B per message). 

With all framing overhead, on-wire messages are ~230B uncompressed vs ~120B compressed. These are the ~real numbers we got from Wireshark experiments.

So, even for messages much smaller than 1024B, compression has significant effect on traffic consumption. In my understanding (might be wrong), traffic is measured by the actual IP level traffic, not rounded up to ethernet packet size.

With that, what do you think about making this at least somehow configurable via the OkHttp API? As a user, I'd be surprised to see that compression for my messages (all of which are under 1K) isn't working, despite being implemented in OkHttp.

**Comment by swankjesse** [comment]
I think we should expose a configuration parameter on OkHttpClient.Builder! Want to send a PR?

**Comment by AntonRutkevich** [comment]
Yep

**Comment by AntonRutkevich** [comment]
#5890

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt [hunk 2]
```diff
@@ -144,6 +144,12 @@ class RealWebSocket(
   }
 
   fun connect(client: OkHttpClient) {
+    if (originalRequest.header("Sec-WebSocket-Extensions") != null) {
+      failWebSocket(ProtocolException(
+          "Request header not permitted: 'Sec-WebSocket-Extensions'"), null)
+      return
```
Comment (sbordet) [inline okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:150]: @swankjesse I think this code is a little too restrictive to applications that want to specify their own extensions or other extensions that are not just the `permessage-deflate`.
I would rather merge the extensions provided by the application with the internal `permessage-deflate`, unless the application already specifies `permessage-deflate`. Thoughts?
FTR I have a CometD test that was testing this that fails with OkHttp 4.5.0. Thanks!

Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:150]: Can you raise a separate issue to track and link to your test case?  Want to make sure this doesn't get lost if we don't immediately fix.

Comment (sbordet) [reply okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:150]: @yschimke filed #5953, thanks!

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt [hunk 3]
```diff
@@ -162,15 +169,24 @@ class RealWebSocket(
...
-          extensions = WebSocketExtensions()
         } catch (e: IOException) {
           exchange?.webSocketUpgradeFailed()
           failWebSocket(e, response)
           response.closeQuietly()
           return
         }
 
+        // Apply the extensions. If they're unacceptable initiate a graceful shut down.
+        // TODO(jwilson): Listeners should get onFailure() instead of onClosing() + onClosed(1010).
```
Comment (nlgtuankiet) [inline okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:180]: @swankjesse  
Is there any reason not to use failWebSocket() here and return?
If not I am happy to submit a pr to fix this TODO

**Pull request merged**